File: src/pl/swmud/ns/swaedit/gui/VnumQuestionWidget.java
package pl.swmud.ns.swaedit.gui;

import java.math.BigInteger;

import com.trolltech.qt.core.Qt;
import com.trolltech.qt.gui.QWidget;

class VnumQuestionWidget extends QWidget {
    
    private Ui_VnumQuestionWidget ui = new Ui_VnumQuestionWidget();
    private BigInteger vnum;
    Signal1<BigInteger> vnumSet = new Signal1<BigInteger>();
    

    public VnumQuestionWidget() {
        ui.setupUi(this);
        SWAEdit.setChildPosition(this);
        setAttribute(Qt.WidgetAttribute.WA_DeleteOnClose);
        setWindowModality(Qt.WindowModality.ApplicationModal);
    }
    
    @SuppressWarnings("unused")
    private void on_vnumEdit_textChanged(String str) {
        try {
            vnum = BigInteger.valueOf(Long.parseLong(str));
            if (vnum.compareTo(BigInteger.ZERO) == 1) {
                ui.acceptButton.setEnabled(true);
            }
        } catch (NumberFormatException e) {
            ui.acceptButton.setEnabled(false);
        }
    }
    
    @SuppressWarnings("unused")
    private void on_acceptButton_clicked() {
        hide();
        vnumSet.emit(vnum);
        close();
    }

    @SuppressWarnings("unused")
    private void on_cancelButton_clicked() {
        close();
    }


}

File: src/gui/VnumQuestionWidget.cpp
#include "gui/VnumQuestionWidget.h"
#include "ui_VnumQuestionWidget.h"

VnumQuestionWidget::VnumQuestionWidget(QWidget *parent)
    : QWidget(parent)
    , ui(new Ui::VnumQuestionWidget)
{
    ui->setupUi(this);
    setWindowFlag(Qt::Window);
    setAttribute(Qt::WA_DeleteOnClose);
    setWindowModality(Qt::ApplicationModal);

    ui->acceptButton->setEnabled(false);

    connect(ui->vnumEdit, &QLineEdit::textChanged,
            this, &VnumQuestionWidget::onVnumEditTextChanged);
    connect(ui->acceptButton, &QPushButton::clicked,
            this, &VnumQuestionWidget::onAcceptClicked);
    connect(ui->cancelButton, &QPushButton::clicked,
            this, &VnumQuestionWidget::onCancelClicked);
}

VnumQuestionWidget::~VnumQuestionWidget()
{
    delete ui;
}

void VnumQuestionWidget::onVnumEditTextChanged(const QString &text)
{
    bool ok;
    qint64 val = text.toLongLong(&ok);
    ui->acceptButton->setEnabled(ok && val > 0);
}

void VnumQuestionWidget::onAcceptClicked()
{
    bool ok;
    qint64 vnum = ui->vnumEdit->text().toLongLong(&ok);
    if (!ok || vnum <= 0) return;

    emit vnumSet(vnum);
    close();
}

void VnumQuestionWidget::onCancelClicked()
{
    close();
}
